Add espeak-ng and pcaudiolib (a dependency of espeak-ng)

egaznep opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io

All apologies in advance if your recipe PR does not recieve prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

egaznep commented:
@conda-forge-admin could you ping @conda-forge/help-c-cpp

egaznep commented:
@conda-forge-admin could you ping @conda-forge/help-c-cpp

bollwyvl commented on recipes/espeak-ng/COPYING:
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
This is included in the upstream tarball, and probably isn't being used.

egaznep replied:
I am still rather a newbie regarding contributing to `conda-forge`. What action should I take then?

bollwyvl replied:
luckily that part's easy: just delete the vendored `COPYING` file, and it see if it builds.

I do see there are bunch of _additional_ `COPYING` files upstream for vendored things, these should be included as well, below in `about`:

```yaml
about:
  #... what's already there, except `license` and `license_file`
  license: GPL-3.0-only
  license_file:
  - COPYING
  - COPYING.APACHE
  - COPYING.BSD2
  - COPYING.UCD
```

egaznep replied:
Thank you, will attend this shortly.

egaznep replied:
@bollwyvl Done, I would really appreciate if you could give it another check

bollwyvl replied:
licensing looks fine. 

It seems like upstream makes a serious effort to support non-Linux platforms, and we should _probably_ do the same.

I can't really provide much feedback on the rest, and especially if you can get it building on `win-64` and `osx-64`, you can request a review from `@conda-forge/help-python-c @conda-forge/help-c-cpp`

carterbox commented on recipes/espeak-ng/meta.yaml:
```diff
@@ -0,0 +1,65 @@
+
+test:
+  commands:
+    - test -f $PREFIX/bin/espeak-ng  # [not win]
```
```suggestion
    - espeak-ng --help
```
```suggestion
    - espeak-ng --version
```
Can you test whether the binary is executable instead?

egaznep replied:
These seem to be working. We could go with any of the two.

carterbox commented on recipes/sctk/0001-fix-compiler-for-rfilter1.patch:
```diff
@@ -1,25 +0,0 @@
-From 0f0f42fa908ac3812cee91645102c440f8f542c7 Mon Sep 17 00:00:00 2001
-From: Unal Ege Gaznepoglu <uenal.ege.gaznepoglu@iis.fraunhofer.de>
```
Do not delete other recipes from the recipes folder. It can cause merge conflicts.

egaznep replied:
This was a recipe that I pushed some time ago. I did not know back then that I could not fork the same repository twice. How should I treat this, by editing the commit history?

carterbox replied:
You can use `git rebase`. Multiple forks of the same repository are not necessary because you can have multiple branches in a fork.

carterbox commented on recipes/pcaudiolib/meta.yaml:
```diff
@@ -0,0 +1,59 @@
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
```
The cxx compiler is only needed for Windows?

carterbox commented on recipes/pcaudiolib/meta.yaml:
```diff
@@ -0,0 +1,59 @@
+    - alsa-lib  # [linux]
+test:
+  commands:
+    - test -f $PREFIX/lib/libpcaudio.so  # [linux]
```
Shared library packages need run_exports, so that downstream packages always depend on a compatible version.

What kind of ABI compatability guarantees and versioning does this project use? Is the ABI guaranteed to only break when the soname major version increases?

carterbox replied:
As a packager of this library, you need to figure it out. Read about conda run_exports and ask the creators of this package what they plan for ABI compatability.

carterbox commented on recipes/pcaudiolib/build.sh:
```diff
@@ -0,0 +1,8 @@
+# custom LD for windows needed
+LD=$LD ./configure --prefix=$(realpath $PREFIX)
+make prefix=$(realpath $PREFIX)
+make install prefix=$(realpath $PREFIX)
```
We don't allow shipping static libraries. Please remove libpcaudio.a as the last build step.

egaznep replied:
Could I do this by `--disable-static` instead?

carterbox replied:
You can if that is an option recognized by the build script.

carterbox commented on recipes/pcaudiolib/meta.yaml:
```diff
@@ -0,0 +1,59 @@
+  commands:
+    - test -f $PREFIX/lib/libpcaudio.so  # [linux]
+    - test -f $PREFIX/lib/libpcaudio.0.dylib  # [osx]
+    - if not exist %PREFIX%\\lib\\pcaudio.lib exit 1  # [win]
```
You should be checking for both pcaudio.dll and pcaudio.lib?

egaznep replied:
As far as I know, the only thing that depends on pcaudiolib is espeak-ng, and on Windows they are linking to the static version. That's why I checked the static one.

carterbox replied:
We don't allow static linking on this channel. On Windows, the .lib extension is used for both static libraries (used for static linking) and import libraries (used for dynamic linking).

carterbox commented on recipes/espeak-ng/meta.yaml:
```diff
@@ -0,0 +1,65 @@
+    - pcaudiolib
+
+test:
+  commands:
```
This package contains a number of other artifacts which should be tested for. e.g. libraries, headers, config files, stuff in $PREFIX/shared

egaznep replied:
In general, what would be the rule of thumb when writing these tests, testing every installed object?

carterbox commented on recipes/espeak-ng/meta.yaml:
```diff
@@ -0,0 +1,65 @@
+  url: https://github.com/espeak-ng/{{ name }}/archive/cb62d93fd7b61d8593b9ae432e6e2a78e3711a77.zip
+  sha256: 81f01cdbc928cefde3989e62ee185285c46659503ca0dc1c81e5de2962afd6ad
+
+build:
+  number: 0
```
```suggestion
build:
  number: 0
  skip: true  # [win]
```
Looks like in the latest dev version, the project has switched from autotools to CMake. Maybe skip Windows until the next release?

egaznep replied:
I tried getting this running on Windows over the weekend, while I figured out it for `pcaudiolib` (it was a total disaster) I still couldn't get it for `espeak-ng`.

traversaro commented:
@egaznep are you still interested in this PR? I may be interested in following it up and addressing reviewers comment, but I wanted to know if that is ok for you before, thanks!

egaznep commented:
> @egaznep are you still interested in this PR? I may be interested in following it up and addressing reviewers comment, but I wanted to know if that is ok for you before, thanks!

@traversaro please feel free to move forward. I have moved away from the project for which I needed this conda-forge recipe, and I won't be able to make time for it in my schedule in the near future. Would I need to do anything such that you can take over?

egaznep commented:
> > > @egaznep are you still interested in this PR? I may be interested in following it up and addressing reviewers comment, but I wanted to know if that is ok for you before, thanks!
> > 
> > 
> > @traversaro please feel free to move forward. I have moved away from the project for which I needed this conda-forge recipe, and I won't be able to make time for it in my schedule in the near future. Would I need to do anything such that you can take over?
> 
> No, I can build on your commits in a separate branch, thanks! By the way I noticed that espeak-ng switched to CMake in its latest release, so I guess the recipe will change a bit anyhow.

I see, I hope the procedure becomes simpler than what I had to deal with. Good luck!
